services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/update/UpdateExpressionConverter.java

```diff
@@ -0,0 +1,188 @@
+     * <p>
+     * Examples: The expression contains a {@link DeleteUpdateAction} with a path value of 'MyAttribute[1]'; the list returned
+     * will have 'MyAttribute' as an element.}
+     *
+     * @return A list of top level attribute names that have update actions associated.
+     */
+    public static List<String> findAttributeNames(UpdateExpression updateExpression) {
+        List<String> attributeNames = listPathsWithoutTokens(updateExpression);
+        List<String> attributeNamesFromTokens = listAttributeNamesFromTokens(updateExpression);
+        attributeNames.addAll(attributeNamesFromTokens);
+        return attributeNames;
+    }
+
+    private static List<String> groupExpressions(UpdateExpression expression) {
+        List<String> groupExpressions = new ArrayList<>();
+        if (!expression.setActions().isEmpty()) {
+            groupExpressions.add(SET + expression.setActions().stream()
+                                                 .map(a -> String.format("%s = %s", a.path(), a.value()))
+                                                 .collect(Collectors.joining(ACTION_SEPARATOR)));
+        }
+        if (!expression.removeActions().isEmpty()) {
+            groupExpressions.add(REMOVE + expression.removeActions().stream()
+                                                    .map(RemoveUpdateAction::path)
+                                                    .collect(Collectors.joining(ACTION_SEPARATOR)));
+        }
+        if (!expression.deleteActions().isEmpty()) {
+            groupExpressions.add(DELETE + expression.deleteActions().stream()
+                                                    .map(a -> String.format("%s %s", a.path(), a.value()))
+                                                    .collect(Collectors.joining(ACTION_SEPARATOR)));
+        }
+        if (!expression.addActions().isEmpty()) {
+            groupExpressions.add(ADD + expression.addActions().stream()
+                                                 .map(a -> String.format("%s %s", a.path(), a.value()))
+                                                 .collect(Collectors.joining(ACTION_SEPARATOR)));
+        }
+        return groupExpressions;
+    }
+
+    private static Stream<Map<String, String>> streamOfExpressionNames(UpdateExpression expression) {
+        return Stream.concat(expression.setActions().stream().map(SetUpdateAction::expressionNames),
+                             Stream.concat(expression.removeActions().stream().map(RemoveUpdateAction::expressionNames),
+                                           Stream.concat(expression.deleteActions().stream()
+                                                                   .map(DeleteUpdateAction::expressionNames),
+                                                         expression.addActions().stream()
+                                                                   .map(AddUpdateAction::expressionNames))));
+    }
```

Comment:
There have been so many opinions on the API and whether there may be instances in the future that wouldn't implement it so I don't quite dare to. Again, since this code is internal we can optimize down the line.

Reply:
Agree it's super clunky though